From akpm@linux-foundation.org  Wed Apr 15 21:31:15 2009
From: Christoph Mair <christoph.mair@gmail.com>
Date: Wed, 15 Apr 2009 16:41:32 -0700
Subject: USB: serial: ti_usb_3410_5052.c: ti_usb returns EIO when reopening the device
To: mm-commits@vger.kernel.org
Cc: christoph.mair@gmail.com, alan@lxorguk.ukuu.org.uk, greg@kroah.com, stable@kernel.org
Message-ID: <200904152341.n3FNfWj2010587@imap1.linux-foundation.org>

From: Christoph Mair <christoph.mair@gmail.com>

Fix regression introduced by 4a90f09b20f4622dcbff1f0e1e6bae1704f8ad8c
("tty: usb-serial krefs").

The driver works as expected until you close the device and try to reopen
it.  All you get from now on, is an I/O error.  It seems that something
isn't released correctly, because the refcount of the usbserial module
does not drop to zero, even when unloading all dependent drivers.

Cc: Alan Cox <alan@lxorguk.ukuu.org.uk>
Cc: <stable@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/serial/ti_usb_3410_5052.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/usb/serial/ti_usb_3410_5052.c
+++ b/drivers/usb/serial/ti_usb_3410_5052.c
@@ -1231,8 +1231,8 @@ static void ti_bulk_in_callback(struct u
 			tport->tp_icount.rx += urb->actual_length;
 			spin_unlock(&tport->tp_lock);
 		}
-		tty_kref_put(tty);
 	}
+	tty_kref_put(tty);
 
 exit:
 	/* continue to read unless stopping */
